cli4/cli4.1
.TH CLI4 1

.SH NAME
cli4 \- Command line access to Cloudflare v4 API

.SH SYNOPSIS
.B cli4
[\fB\-V\fR|\fB\-\-version]
[\fB\-h\fR|\fB\-\-help]
[\fB\-v\fR|\fB\-\-verbose]
[\fB\-e\fR|\fB\-\-examples]
[\fB\-q\fR|\fB\-\-quiet]
[\fB\-j\fR|\fB\-\-json]
[\fB\-y\fR|\fB\-\-yaml]
[\fB\-n\fR|\fB\-\-ndjson]
[\fB\-r\fR|\fB\-\-raw]
[\fB\ir\fR|\fB\-\-image]
[\fB\-d\fR|\fB\-\-dump]
[\fB\-b\fR|\fB\-\-binary]
[\fB\-A openapi-url\fR|\fB\-\-openapi openapi-url]
[\fB\-p profile-name\fR|\fB\-\-profile profile-name]
[\fBitem\fR=\fIvalue\fR ...]
[=\fIvalue\fR ...]
[\fBitem\fR=@\fIfilename\fR ...]
[=@\fIfilename\fR ...]
[\fB\-G\fR|\fB\-\-get]
[\fB\-P\fR|\fB\-\-patch]
[\fB\-O\fR|\fB\-\-post]
[\fB\-U\fR|\fB\-\-put]
[\fB\-D\fR|\fB\-\-delete]
.IR /command ...

.SH DESCRIPTION
.B cli4
provides command line access to Cloudflare v4 API

.SH OPTIONS
.TP
.IP "[\-V, \-\-version]"
Display program version number and exit.
.IP "[\-h, \-\-help]"
This information (in a terse form).
.IP "[\-v, \-\-verbose]"
Provide some protocol debugging information.
.IP "[\-e, \-\-example]"
Show the path to the examples folder/directory.
.IP "[\-q, \-\-quiet]"
Don't output any JSON/YAML responses.
.IP "[\-j, \-\-json]"
Output response data in JSON format (the default).
.IP "[\-y, \-\-yaml]"
Output response data in YAML format (if yaml package installed).
.IP "[\-n, \-\-ndjson]"
Output response data in NDJSON format (if jsonlines package installed).
.IP "[\-r, \-\-raw]"
Output JSON results in raw mode without splitting out the errors and results.
.IP "[\-i, \-\-image]"
Output results as binary (i.e. it's an image).
This isn't normally needed as the API returns a Content-Type to indicate an image is being returned.
This option can be used to allow the API output to be output without any processing (i.e. JSON formatted).
.IP "[\-d, \-\-dump]"
Output a list of all API calls included in the code.
.IP "[\-b, \-\-binary]"
Open files in binary mode.
.IP "[-A \fIopenapi-url\fR, \-\-openapi \fIopenapi-url\fR]"
Decode Cloudflare's OpenAPI spec and output all the known commands.
.IP "[-p \fIprofile-name\fR, \-\-profile \fIprofile-name\fR]"
Select a \fIprofile-name\fR from the configuration file (hence select custom \fIemail\fR/\fItoken\fR values).
.IP "\-\-get"
Send HTTP request as a \fBGET\fR (the default).
.IP "\-\-patch"
Send HTTP request as a \fBPATCH\fR.
.IP "\-\-post"
Send HTTP request as a \fBPOST\fR.
.IP "\-\-put"
Send HTTP request as a \fBPUT\fR.
.IP "\-\-delete"
Send HTTP request as a \fBDELETE\fR.
.IP "item=\fIvalue\fR"
Set a parameter or data value to send with a \fBGET\fR, \fBPATCH\fR, \fBPOST\fR, \fBPUT\fR or \fBDELETE\fR command. The value is sent as a string.
.IP item:=\fIvalue\fR
Set a parameter or data value to send with a \fBGET\fR, \fBPATCH\fR, \fBPOST\fR, \fBPUT\fR or \fBDELETE\fR command. The value is sent as an integer.
.IP item=@\fIfilename\fR
Set a parameter or data value to send with a \fBPOST\fR or \fBPUT\fR command. The value is based on the content of the file.
.IP "\fI/command ...\fR"
The API command(s) to execute.

.SH COMMAND(S)
The command string uses slash (\fB/\fR) to separate the verbs in the same way that the Cloudflare v4 API documentation does.
Any verb starting with colon (\fB:\fR) is either converted to zone_id, user_id, organtization_id, or otherwise.
Any verb starting with coloe (\fB:\fR) and followed by 32 hex characters is passed thru raw.
Any verb starting with two colons (\fB::\fR) is passed thru raw.

.SH RESULTS
The output is either JSON or YAML formatted.

.SH EXAMPLES
.B cli4 /zones
List information for all zones.

.B cli4 /zones/:example.com
List specific zone info.

.B cli4 /zones/:example.com/settings
List settings for a specific zone.

.B cli4 --delete purge_everything=true /zones/:example.com/purge_cache
Purge cache for a specific zone.

.B cli4 --delete files='[http://example.com/css/styles.css]' /zones/:example.com/purge_cache
Purge cache for a specific zone.

.B cli4 --delete files='[http://example.com/css/styles.css,http://example.com/js/script.js] /zones/:example.com/purge_cache
Purge cache for a specific zone.

.B cli4 --delete tags='[tag1,tag2,tag3]' /zones/:example.com/purge_cache
Purge cache for a specific zone.

.B cli4 /zones/:example.com/available_plans
List available plans for a zone.

.B cli4 --patch status=active /zones/:example.com/dnssec
Make DNSSEC active for specific zone.

.B cli4 /zones/:example.com/dnssec
List DNSSEC information and status for a specific zone.

.SH SEE ALSO
The Cloudflare API can be found https://api.cloudflare.com/. Each API call is provided via a similarly named function within the Cloudflare class.


CHANGELOG.md
# Change Log

 - 2024-04-29 18:38:21 +0100 [1bda322](../../commit/1bda322ccb2093c262cb65777327b1d1133d2a5a) 2.19.4
 - 2024-04-29 18:36:37 +0100 [a3cda63](../../commit/a3cda63d40e9bf25cbd90bef76b28ad16285c2ba) #186 - explain how to maps arguments
 - 2024-04-29 18:28:56 +0100 [389029a](../../commit/389029a6c8c20cd125333c81f758ffb718169ab9) #188 - url now uses strings - as it should!
 - 2024-04-29 18:07:15 +0100 [4bd76db](../../commit/4bd76dbd42a84a851719b7a5a3f47176db3a189f) #190 - remove Python 3.5 support because of f-string
 - 2024-04-29 18:04:12 +0100 [02e513f](../../commit/02e513f1607be04f031ee9387872d6666cabeb1a) 2.19.3
 - 2024-04-29 18:02:44 +0100 [8f2e7ad](../../commit/8f2e7ad974c71a4b507eb90ddc8d0210f318ab4e) Add version 3.0 code and pinning info to README
 - 2024-04-29 18:01:58 +0100 [b4dbc6e](../../commit/b4dbc6e8ee954ae264f72e2ac38d4f9c47201e19) more api endpoints
 - 2024-04-29 18:01:25 +0100 [63f1839](../../commit/63f18392d3346924f81e59e681a877a940cbd656) now that AI methods are in library - no need for CLOUDFLARE_API_EXTRAS code
 - 2024-04-18 16:35:32 -0700 [0c1fb92](../../commit/0c1fb926b0bf0595b025272b9e80bc77ed7c5ad3) more api endpoints
 - 2024-03-19 15:02:14 -0700 [839ac32](../../commit/839ac3206320206178fd67438a52e0f4357f3903) remove f-string for Python <3.6 compatibility
 - 2024-03-11 22:25:20 -0300 [ab8a1aa](../../commit/ab8a1aaa189ac5afdab20f966ef44fd675e1437b) added documentation urls, etc for pypi
 - 2024-03-11 22:21:01 -0300 [fc0e111](../../commit/fc0e1111e1824ea8c78cda43530b51343fb0a348) finally removed the VOID calls, added /ai/run endpoints
 - 2024-03-04 19:50:56 -0500 [56ffcfc](../../commit/56ffcfc1557610ae4d18eb305fbeb418966f038e) still testing docs
 - 2024-03-04 19:46:21 -0500 [82d2dcb](../../commit/82d2dcb2f8dc51bc72e3bca455bba38c55e66f8b) added more download stats
 - 2024-03-04 19:32:03 -0500 [1ae9c2a](../../commit/1ae9c2aa7a67f8f15bce6cdabc3a2b71a8744c11) still testing docs
 - 2024-03-04 19:29:14 -0500 [f9b565a](../../commit/f9b565a772c1ff06b2ed8a432734fa67ae2b24bc) added download stats
 - 2024-03-04 00:02:19 -0500 [984284f](../../commit/984284f563e90c91e296a0c5d5d2f65dad9f8a4e) still testing docs
 - 2024-03-04 00:01:25 -0500 [cf2b3d7](../../commit/cf2b3d7b6630f677e9224ed91ea4b8d462fbdd92) still testing docs
 - 2024-03-01 12:05:09 -0800 [26f6af8](../../commit/26f6af8bc79ab6c3c8c0c8526a73012d46350c0d) still testing docs
 - 2024-03-01 11:39:45 -0800 [35c0e06](../../commit/35c0e068df3459f6cfcb4c39cadd15f526386e4c) start of longstanding need for documentation, update of copyright string
 - 2024-03-01 10:15:38 -0800 [4fe03ee](../../commit/4fe03eed3c7bcc2130846b1be6ea4c9427a1bb99) readthedocs initial setup - long overdue
 - 2024-02-26 23:33:35 -0800 [77fee2e](../../commit/77fee2e5fcf9e3246f6f2d9f194dbb65e54cdcf4) pylint fixes
 - 2024-02-26 23:30:57 -0800 [e7cc964](../../commit/e7cc96452fb84338525ed1537e3523157ba634de) moved all requests code/exceptions into network where it belongs, import cleanup, exception handling cleanup
 - 2024-02-26 23:20:06 -0800 [dbaf7d6](../../commit/dbaf7d67c53e0baaf7a1ab6f33ee860c50672686) handle python < 3.10 to get utc time correctly
 - 2024-02-23 16:47:01 -0800 [036c5ed](../../commit/036c5edeac4d40e8bd9584541ce4a84f2e686d20) document cli4 having --header flag
 - 2024-02-23 16:39:32 -0800 [6eed220](../../commit/6eed2201254769b9bbcea69252fdad51bae62aa2) cli4 has --header flag, plus http_headers needed some more syntax checking
 - 2024-02-22 19:03:24 -0800 [dec9835](../../commit/dec98352d027e205bed762bebdbc8ea0eecbd7c5) added Python 3.5 note
 - 2024-02-22 15:10:24 -0800 [b3cf637](../../commit/b3cf637c2318e6aa09150a3f5f55e5a8a90221f3) remove excessive import statements
 - 2024-02-22 15:09:41 -0800 [c5a5587](../../commit/c5a5587abff4213ffc9fdeac515390a0982f73ee) improved date/time rfc/iso code, lint fixes
 - 2024-02-22 13:49:22 +0900 [9dca32b](../../commit/9dca32b2eb2511ce7b5ae3585ffec17b0eab8278) CHANGELOG.md pushed to github
 - 2024-02-22 13:49:15 +0900 [28b768a](../../commit/28b768a0d02309cef9f150bfc41cb9c3d9c2eecd) 2.19.2
 - 2024-02-22 13:48:52 +0900 [11d1270](../../commit/11d1270b65ce089b87b029483f517c56d288bff5) typo
 - 2024-02-22 13:37:13 +0900 [287c8a7](../../commit/287c8a790b3adf80352972d9ca63c14eab2f8c3a) CHANGELOG.md pushed to github
 - 2024-02-22 13:36:59 +0900 [6f9e5cf](../../commit/6f9e5cf637fb743e6e9a231ac2b3e37211d492d4) 2.19.1
 - 2024-02-22 13:34:05 +0900 [08123c1](../../commit/08123c1cb1fd49c5727c89d5bfdac95f36f01624) http_headers documentation
 - 2024-02-22 13:14:37 +0900 [44751a0](../../commit/44751a003c33d75d73d51889728e75be4bafb7f5) fix tag name creation
 - 2024-02-22 11:25:20 +0900 [10d8af3](../../commit/10d8af3dce16af13127cb3e7fe2688acbdca9ad0) CHANGELOG.md pushed to github

cli4/cli4.py
#!/usr/bin/env python
"""Cloudflare API via command line"""

import sys
import re
import getopt
import json

import CloudFlare

from .dump import dump_commands, dump_commands_from_web
from . import converters
from . import examples

my_yaml = None
my_jsonlines = None

class CLI4InternalError(Exception):
    """ errors in cli4 """

def load_and_check_yaml():
    """ load_and_check_yaml() """
    # only called if user uses --yaml flag
    from . import myyaml
    global my_yaml
    try:
        my_yaml = myyaml.myyaml()
    except ImportError:
        sys.exit('cli4: install yaml support via: pip install pyyaml')

def load_and_check_jsonlines():
    """ load_and_check_yaml() """
    # only called if user uses --ndjson flag
    from . import myjsonlines
    global my_jsonlines
    try:
        my_jsonlines = myjsonlines.myjsonlines()
    except ImportError:
        sys.exit('cli4: install jsonlines support via: pip install jsonlines')

def strip_multiline(s):
    """ remove leading/trailing tabs/spaces on each line"""
    # This hack is needed in order to use yaml.safe_load() on JSON text - tabs are not allowed
    return '\n'.join([line.strip() for line in s.splitlines()])

def process_params_content_files(method, binary_file, args):
    """ process_params_content_files() """

    digits_only = re.compile('^-?[0-9]+$')
    floats_only = re.compile('^-?[0-9.]+$')

    params = None
    content = None
    files = None
    # next grab the params. These are in the form of tag=value or =value or @filename
    while len(args) > 0 and ('=' in args[0] or args[0][0] == '@'):
        arg = args.pop(0)
        if arg[0] == '@':
            # a file to be uploaded - used in workers/script etc - only via PUT or POST
            filename = arg[1:]
            if method not in ['PUT','POST']:
                sys.exit('cli4: %s - raw file upload only with PUT or POST' % (filename))
            try:
                if filename == '-':
                    if binary_file:
                        content = sys.stdin.buffer.read()
                    else:
                        content = sys.stdin.read()
                else:
                    if binary_file:
                        with open(filename, 'rb') as f:
                            content = f.read()
                    else:
                        with open(filename, 'r', encoding="utf-8") as f:
                            content = f.read()
            except IOError:
                sys.exit('cli4: %s - file open failure' % (filename))
            continue
        tag_string, value_string = arg.split('=', 1)
        if value_string.lower() == 'true':
            value = True
        elif value_string.lower() == 'false':
            value = False
        elif value_string == '' or value_string.lower() == 'none':
            value = None
        elif value_string[0] == '=' and value_string[1:] == '':
            sys.exit('cli4: %s== - no number value passed' % (tag_string))
        elif value_string[0] == '=' and digits_only.match(value_string[1:]):
            value = int(value_string[1:])
        elif value_string[0] == '=' and floats_only.match(value_string[1:]):
            value = float(value_string[1:])
        elif value_string[0] == '=':
            sys.exit('cli4: %s== - invalid number value passed' % (tag_string))
        elif value_string[0] in '[{' and value_string[-1] in '}]':
            # a json structure - used in pagerules
            try:
                # value = json.loads(value) - changed to yaml code to remove unicode string issues
                load_and_check_yaml()
                # cleanup string before parsing so that yaml.safe.load does not complain about whitespace
                # >>> found character '\t' that cannot start any token <<<
                value_string = strip_multiline(value_string)
                try:
                    value = my_yaml.safe_load(value_string)
                except my_yaml.parser.ParserError:
                    raise ValueError from None
            except ValueError:
                sys.exit('cli4: %s="%s" - can\'t parse json value' % (tag_string, value_string))
        elif value_string[0] == '@':
            # a file to be uploaded - used in dns_records/import etc - only via PUT or POST
            filename = value_string[1:]
            if method not in ['PUT', 'POST']:
                sys.exit('cli4: %s=%s - file upload only with PUT or POST' % (tag_string, filename))
            if files is None:
                files = {}
            if tag_string in files:
                sys.exit('cli4: %s=%s - duplicate name' % (tag_string, filename))
            try:
                if filename == '-':
                    files[tag_string] = sys.stdin
                else:
                    files[tag_string] = open(filename, 'rb')
            except IOError:
                sys.exit('cli4: %s=%s - file open failure' % (tag_string, filename))
            # no need for param code below
            continue
        elif (value_string[0] == '"' and value_string[-1] == '"') or (value_string[0] == '\'' and value_string[-1] == '\''):
            # remove quotes
            value = value_string[1:-1]
        else:
            value = value_string

        if tag_string == '':
            # There's no tag; it's just an unnamed list
            if params is None:
                params = value
            else:
                sys.exit('cli4: %s=%s - param error. Can\'t mix unnamed and named list' %
                         (tag_string, value_string))
        else:
            if params is None:
                params = {}
            tag = tag_string
            try:
                params[tag] = value
            except TypeError:
                sys.exit('cli4: %s=%s - param error. Can\'t mix unnamed and named list' %
                         (tag_string, value_string))

    if content and params:
        sys.exit('cli4: content and params not allowed together')

    if params and files:
        for k,v in params.items():
            files[k] = (None, v)
        params = None
        # sys.exit('cli4: params and files not allowed together')

    if method != 'GET':
        if params:
            content = params
            params = None

    return (params, content, files)

def run_command(cf, method, command, params=None, content=None, files=None):
    """run the command line"""
    # remove leading and trailing /'s
    if command[0] == '/':
        command = command[1:]
    if command[-1] == '/':
        command = command[:-1]

    # break down command into it's separate pieces
    # these are then checked against the Cloudflare class
    # to confirm there is a method that matches
    parts = command.split('/')

    cmd = []
    identifier1 = None
    identifier2 = None
    identifier3 = None

    hex_only = re.compile('^[0-9a-fA-F]+$')
    waf_rules = re.compile('^[0-9]+[A-Z]*$')
    uuid_value = re.compile('^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}$') # 8-4-4-4-12

    m = cf
    for element in parts:
        if element[0] == ':':
            element = element[1:]
            if identifier1 is None:
                if len(element) in [32, 40, 48] and hex_only.match(element):
                    # raw identifier - lets just use it as-is
                    identifier1 = element
                elif len(element) == 36 and uuid_value.match(element):
                    # uuid identifier - lets just use it as-is
                    identifier1 = element
                elif element[0] == ':':
                    # raw string - used for workers script_name - use ::script_name
                    identifier1 = element[1:]
                else:
                    try:
                        if cmd[0] == 'certificates':
                            # identifier1 = convert_certificates_to_identifier(cf, element)
                            identifier1 = converters.convert_zones_to_identifier(cf, element)
                        elif cmd[0] == 'zones':
                            identifier1 = converters.convert_zones_to_identifier(cf, element)
                        elif cmd[0] == 'accounts':
                            identifier1 = converters.convert_accounts_to_identifier(cf, element)
                        elif cmd[0] == 'organizations':
                            identifier1 = converters.convert_organizations_to_identifier(cf, element)
                        elif (cmd[0] == 'user') and (cmd[1] == 'organizations'):
                            identifier1 = converters.convert_organizations_to_identifier(cf, element)
                        elif (cmd[0] == 'user') and (cmd[1] == 'invites'):
                            identifier1 = converters.convert_invites_to_identifier(cf, element)
                        elif (cmd[0] == 'user') and (cmd[1] == 'virtual_dns'):
                            identifier1 = converters.convert_virtual_dns_to_identifier(cf, element)
                        elif (cmd[0] == 'user') and (cmd[1] == 'load_balancers') and (cmd[2] == 'pools'):
                            identifier1 = converters.convert_load_balancers_pool_to_identifier(cf, element)
                        else:
                            raise CLI4InternalError("/%s/%s :NOT CODED YET" % ('/'.join(cmd), element))
                    except CLI4InternalError as e:
                        sys.stderr.write('cli4: /%s - %s\n' % (command, e))
                        raise e
                cmd.append(':' + identifier1)
            elif identifier2 is None:
                if len(element) in [32, 40, 48] and hex_only.match(element):
                    # raw identifier - lets just use it as-is
                    identifier2 = element
                elif len(element) == 36 and uuid_value.match(element):
                    # uuid identifier - lets just use it as-is
                    identifier2 = element
                elif element[0] == ':':
                    # raw string - used for workers script_names
                    identifier2 = element[1:]
                else:
                    try:
                        if (cmd[0] and cmd[0] == 'zones') and (cmd[2] and cmd[2] == 'dns_records'):
                            identifier2 = converters.convert_dns_record_to_identifier(cf, identifier1, element)
                        elif (cmd[0] and cmd[0] == 'zones') and (cmd[2] and cmd[2] == 'custom_hostnames'):
                            identifier2 = converters.convert_custom_hostnames_to_identifier(cf, identifier1, element)
                        else:
                            raise CLI4InternalError("/%s/:%s :NOT CODED YET" % ('/'.join(cmd), element))
                    except CLI4InternalError as e:
                        sys.stderr.write('cli4: /%s - %s\n' % (command, e))
                        raise e
                # identifier2 may be an array - this needs to be dealt with later
                if isinstance(identifier2, list):
                    cmd.append(':' + '[' + ','.join(identifier2) + ']')
                else:
                    cmd.append(':' + identifier2)
                    identifier2 = [identifier2]
            else:
                if len(element) in [32, 40, 48] and hex_only.match(element):
                    # raw identifier - lets just use it as-is
                    identifier3 = element
                elif len(element) == 36 and uuid_value.match(element):
                    # uuid identifier - lets just use it as-is
                    identifier3 = element
                elif waf_rules.match(element):
                    identifier3 = element
                elif element[0] == ':':
                    # raw string - used for workers script_names
                    identifier3 = element[1:]
                else:
                    #  /accounts/:id/storage/kv/namespaces/:id/values/:key_name - it's a strange one!
                    if len(cmd) >= 6 and cmd[0] == 'accounts' and cmd[2] == 'storage' and cmd[3] == 'kv' and cmd[4] == 'namespaces' and cmd[6] == 'values':
                        identifier3 = element
                    else:
                        sys.stderr.write('/%s/:%s :NOT CODED YET\n' % ('/'.join(cmd), element))
                        raise e
        else:
            try:
                m = getattr(m, CloudFlare.CloudFlare.sanitize_verb(element))
                cmd.append(element)
            except AttributeError as e:
                # the verb/element was not found
                sys.stderr.write('cli4: /%s - not found\n' % (command))
                raise e

    results = []
    if identifier2 is None:
        identifier2 = [None]
    for i2 in identifier2:
        try:
            if method == 'GET':
                # no content with a GET call
                r = m.get(identifier1=identifier1,
                          identifier2=i2,
                          identifier3=identifier3,
                          params=params)
            elif method == 'PATCH':
                r = m.patch(identifier1=identifier1,
                            identifier2=i2,
                            identifier3=identifier3,
                            data=content)
            elif method == 'POST':
                r = m.post(identifier1=identifier1,
                           identifier2=i2,
                           identifier3=identifier3,
                           data=content, files=files)
            elif method == 'PUT':
                r = m.put(identifier1=identifier1,
                          identifier2=i2,
                          identifier3=identifier3,
                          data=content, files=files)
            elif method == 'DELETE':
                r = m.delete(identifier1=identifier1,
                             identifier2=i2,
                             identifier3=identifier3,
                             data=content)
            else:
                pass
        except CloudFlare.exceptions.CloudFlareAPIError as e:
            if len(e) > 0:
                # more than one error returned by the API
                for x in e:
                    sys.stderr.write('cli4: /%s - %d %s\n' % (command, int(x), str(x)))
            sys.stderr.write('cli4: /%s - %d %s\n' % (command, int(e), str(e)))
            raise e
        except CloudFlare.exceptions.CloudFlareInternalError as e:
            sys.stderr.write('cli4: InternalError: /%s - %d %s\n' % (command, int(e), str(e)))
            raise e
        except Exception as e:
            sys.stderr.write('cli4: /%s - %s - api error\n' % (command, str(e)))
            raise e

        results.append(r)
    return results

def write_results(results, output):
    """dump the results"""

    if output is None:
        return

    if len(results) == 1:
        results = results[0]

    if isinstance(results, (str, bytes, bytearray)):
        # if the results are a simple string, then it should be dumped directly
        # this is only used for /zones/:id/dns_records/export, workers, and other calls
        # or
        # output is image or audio or video or something like that so we dump directly
        pass
    else:
        # anything more complex (dict, list, etc) should be dumped as JSON/YAML
        if output == 'json':
            try:
                results = json.dumps(results,
                                     indent=4,
                                     sort_keys=True,
                                     ensure_ascii=False,
                                     encoding='utf8')
            except TypeError:
                results = json.dumps(results,
                                     indent=4,
                                     sort_keys=True,
                                     ensure_ascii=False)
        elif output == 'yaml':
            results = my_yaml.safe_dump(results)
        elif output == 'ndjson':
            # NDJSON support seems like a hack. There has to be a better way
            try:
                writer = my_jsonlines.Writer(sys.stdout)
                writer.write_all(results)
                writer.close()
            except (BrokenPipeError, IOError):
                pass
            return
        else:
            # None of the above, so pass thru results except something in byte form
            if not isinstance(results, (bytes, bytearray)):
                results = str(results)

    if results:
        try:
            if isinstance(results, (bytes, bytearray)):
                sys.stdout.buffer.write(results)
            else:
                sys.stdout.write(results)
                if not results.endswith('\n'):
                    sys.stdout.write('\n')
        except (BrokenPipeError, IOError):
            pass

def do_it(args):
    """Cloudflare API via command line"""

    verbose = False
    output = 'json'
    example = False
    raw = False
    do_dump = False
    do_openapi = None
    openapi_url = None
    binary_file = False
    profile = None
    http_headers = None
    method = 'GET'

    usage = ('usage: cli4 '
             + '[-V|--version] [-h|--help] [-v|--verbose] '
             + '[-e|--examples] '
             + '[-q|--quiet] '
             + '[-j|--json] [-y|--yaml] [-n|--ndjson] [-i|--image] '
             + '[-r|--raw] '
             + '[-d|--dump] '
             + '[-A|--openapi url] '
             + '[-b|--binary] '
             + '[-p|--profile profile-name] '
             + '[-h|--header additional-header] '
             + '[--get|--patch|--post|--put|--delete] '
             + '[item=value|item=@filename|@filename ...] '
             + '/command ...')

    try:
        opts, args = getopt.getopt(args,
                                   'VhveqjynirdA:bp:h:GPOUD',
                                   [
                                       'version', 'help', 'verbose',
                                       'examples',
                                       'quiet',
                                       'json', 'yaml', 'ndjson', 'image',
                                       'raw',
                                       'dump',
                                       'openapi=',
                                       'binary',
                                       'profile=',
                                       'header=',
                                       'get', 'patch', 'post', 'put', 'delete'
                                   ])
    except getopt.GetoptError:
        sys.exit(usage)
    for opt, arg in opts:
        if opt in ('-V', '--version'):
            sys.exit('Cloudflare library version: %s' % (CloudFlare.__version__))
        if opt in ('-h', '--help'):
            sys.exit(usage)
        elif opt in ('-v', '--verbose'):
            verbose = True
        elif opt in ('-q', '--quiet'):
            output = None
        elif opt in ('-e', '--examples'):
            example = True
        elif opt in ('-j', '--json'):
            output = 'json'
        elif opt in ('-y', '--yaml'):
            load_and_check_yaml()
            output = 'yaml'
        elif opt in ('-n', '--ndjson'):
            load_and_check_jsonlines()
            output = 'ndjson'
        elif opt in ('-i', '--image'):
            output = 'image'
        elif opt in ('-r', '--raw'):
            raw = True
        elif opt in ('-p', '--profile'):
            profile = arg
        elif opt in ('-h', '--header'):
            if http_headers is None:
                http_headers = []
            http_headers.append(arg)
        elif opt in ('-d', '--dump'):
            do_dump = True
        elif opt in ('-A', '--openapi'):
            do_openapi = True
            openapi_url = arg if arg != '' else None
        elif opt in ('-b', '--binary'):
            binary_file = True
        elif opt in ('-G', '--get'):
            method = 'GET'
        elif opt in ('-P', '--patch'):
            method = 'PATCH'
        elif opt in ('-O', '--post'):
            method = 'POST'
        elif opt in ('-U', '--put'):
            method = 'PUT'
        elif opt in ('-D', '--delete'):
            method = 'DELETE'

    if example:
        try:
            examples.display()
        except ModuleNotFoundError as e:
            sys.exit(e)
        sys.exit(0)

    try:
        cf = CloudFlare.CloudFlare(debug=verbose, raw=raw, profile=profile, http_headers=http_headers)
    except Exception as e:
        sys.exit(e)

    if do_dump:
        a = dump_commands(cf)
        # success - just dump results and exit
        sys.stdout.write(a)
        sys.exit(0)

    if do_openapi:
        try:
            a = dump_commands_from_web(cf, openapi_url)
        except CloudFlare.exceptions.CloudFlareAPIError as e:
            sys.exit('cli4: %s - Failed' % (e))
        # success - just dump results and exit
        sys.stdout.write(a)
        sys.exit(0)

    # next grab the params. These are in the form of tag=value or =value or @filename
    (params, content, files) = process_params_content_files(method, binary_file, args)

    # what's left is the command itself
    if len(args) < 1:
        sys.exit(usage)
    commands = args

    exit_with_error = False
    for command in commands:
        try:
            results = run_command(cf, method, command, params, content, files)
            write_results(results, output)
        except KeyboardInterrupt as e:
            sys.exit('cli4: %s - Interrupted\n' % (command))
        except Exception as e:
            exit_with_error = True

    if exit_with_error:
        sys.exit(1)

def cli4(args):
    """Cloudflare API via command line"""

    do_it(args)
    sys.exit(0)
